src/rulesets/Base/fastmath_able.jl

```diff
@@ -36,10 +36,86 @@ let
 
 
         # Unary complex functions
-        @scalar_rule abs(x::Real) sign(x)
-        @scalar_rule abs2(x) 2x
-        @scalar_rule angle(x::Real) Zero()
-        @scalar_rule conj(x::Real) One()
+        ## abs
+        function frule((_, Δx), ::typeof(abs), x::Real)
+            return abs(x), sign(x) * real(Δx)
+        end
+        function frule((_, Δz), ::typeof(abs), z::Complex)
+            Ω = abs(z)
+            return Ω, (real(z) * real(Δz) + imag(z) * imag(Δz)) / Ω
+        end
+        
+        function rrule(::typeof(abs), x::Real)
+            function abs_pullback(Δx)
+                return (NO_FIELDS, real(Δx)*sign(x))
+            end
+            return abs(x), abs_pullback
+        end
+        function rrule(::typeof(abs), z::Complex)
+            Ω = abs(z)
+            function abs_pullback(Δz)
+                return (NO_FIELDS, real(Δz)*z/Ω)
+            end
+            return Ω, abs_pullback
+        end
+
+        ## abs2
+        function frule((_, Δx), ::typeof(abs2), x::Real)
+            return abs2(x), 2x * real(Δx)
+        end
+        function frule((_, Δz), ::typeof(abs2), z::Complex)
+            return abs2(z), 2 * (real(z) * real(Δz) + imag(z) * imag(Δz))
+        end
```

Comment:
How about merging these two methods to 
```suggestion
        function frule((_, Δz), ::typeof(abs2), z)
            return abs2(z), 2 * (real(z) * real(Δz) + imag(z) * imag(Δz))
        end
```
I would say we should only treat the real case separately if doing so allows for some performance optimisations (as it does e.g. for `abs` rules).

Reply:
For `Float64`, I think something funny happens that lets it do this in only one clock cycle:
```julia
let Δz = Ref(1.0 + im), x = Ref(4.0) 
    @btime f1($Δz[], $x[])
    @btime f2($Δz[], $x[])
end

#+RESULTS:
:   1.329 ns (0 allocations: 0 bytes)
:   1.329 ns (0 allocations: 0 bytes)
: 8.0
```
But generally, since the compiler can't know that `imag(Δz)` isn't `Inf` or `NaN`, I think the specialized rules for `Real` will generally be more performant. But yeah, it's also not a big deal, so if it's too much extra cruft we can remove it.

Reply:
Ah, I see. Thanks for your explanation! I think we should keep your version, then.